### PR TITLE
[WIP] OP_BRIBEVERIFY implementation

### PR DESCRIPTION
**Pull request by Christewart**
This pull request implements OP_BRIBEVERIFY. This new op code is intended to be soft forked in to bitcoin core. This is a work in progress, feel free to contribute. There is discussion of how the op code should be implemented on the mailing list here: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-May/014408.html

### CODE REVIEW

**Comment by CryptAxe** [comment]
Looking great so far, thanks for fixing the way OP_BRIBEVERIFY errors out.

**Comment by Christewart** [comment]
If anyone is more familar with automake I would *really* appreciate some one to look at why bitcoin-tx is failing and how to fix it. I *think* it is because [chainActive](https://github.com/Christewart/bitcoin/blob/c355e39fbe2ea48856ea86b25cb8a97710feb032/src/script/interpreter.cpp#L1388) is used for the first time on the consensus rule set. Therefore [`validation.h` where chainActive is defined](https://github.com/Christewart/bitcoin/blob/op_bribe_implementation/src/validation.h#L487) needs to be added somewhere new in the Make file. I'm really unsure of how to do this in an efficient way though. I would appreciate some insight/help.

**Comment by CryptAxe** [comment]
Hey I'm not sure if my email went through to bitcoin-dev yet but I'm pretty
sure we can drop the sidechain id from bribes and do everything without it.

As for the rest I'm looking into it. I need to move some of the drivechain
code around anyways (thanks for pointing out errors my system didn't have,
travis) and I'll see about this as well.

**Comment by Christewart** [comment]
Note: I changed the name of `OP_BRIBEVERIFY` back to `OP_NOP4`. I'm pretty sure bitcoin core won't allow these names to be changed until the new op is successfully soft forked in.

**Comment by Christewart** [comment]
Here is a draft BIP on the work I've done so far, would appreciate feedback before I post to the bitcoin-dev mailing list: https://github.com/Christewart/bips/blob/master/bip-op-bribeverify.mediawiki

**Comment by CryptAxe** [comment]
@Christewart I just finished fixing all of the build errors in the mainchainBMM repository and triggered a rebuild of this PR. We'll see what errors travis finds now and go from there.

**Comment by CryptAxe** [comment]
This will have to be rebased first actually.

**Comment by Christewart** [comment]
Hi @CryptAxe 

I'll rebase today and push. I'm also going to review your pull request and see if I have questions or find possible bugs. 

Don't merge this though even if travis ci succeeds -- definitely still a WIP. Here is the mailing list discussion on the BIP if you haven't seen it: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-June/014663.html

**Comment by Christewart** [comment]
@CryptAxe May I suggest you start working on a draft BIP for the work you've done so far? It seems a lot of people want to start understanding drivechains, but documentation seems to be a common problem.

**Comment by Christewart** [comment]
I've updated the BIP with the changes made in 2a51ee9 and 6a00103. 

https://github.com/Christewart/bips/blob/master/bip-op-bribeverify.mediawiki

The major change is adding the sidechain id to the coinbase commitment structure. This is needed if users decide to primarily use lightning channels for OP_BV scripts -- otherwise we wouldn't have a way to detect if we have multiple commitments in one block for a drivechain.

**Comment by CryptAxe** [comment]
@Christewart Sorry for the delayed response! I've been trying to fix the WT^ updating system in line with suggestions from Paul. I've dropped all of the nasty state script code in the PR that I just opened (https://github.com/drivechain-project/bitcoin/pull/18). 


Thank you for the Python tests! I noticed a TODO comment about generating a random hash, which I believe you can do this way for a fake psuedo-random sha256:

```
import random
hash = random.getrandbits(256)
```

That might be wrong though, I'm not very experienced with Python beyond playing around with it. There might even be a reimplementation of the GetRandHash() function from the Boost Bitcoin unit tests somewhere in Python.

I definitely agree with the ``` SidechainNumberValid -> IsSidechainNumberValid ``` rename. 

As far as this being safe:
https://github.com/Christewart/bitcoin/blob/6a00103d37942966da648664fa613397f3ad012e/src/sidechain.cpp#L31-L32

Since we do know the type we are casting (CSidechainNumber::getint() returns an int) I believe that a static cast is safe in this case. However, there are a few C++ developers with years more experience than myself in this space and we might as well be safe and get one of them to look at this code.

I haven't had time to look at everything yet, but looking good. I need to write up some info about the new WT^ update process (mainly so that it makes sense to you, or if anyone else happens to try and help) and then maybe the three of us (@psztorc being the third person working on Drivechains) should all have an irc / slack meetup or something to organize ourselves and slowly start the process of moving your new code and my new code into the project.

**Comment by Christewart** [comment]
I agree with the sentiment about organizing a meeting of some sort. I think we should do it on freenode since that is where the rest of the dev community coalesces. A logical place is the sidechains-dev channel. If you want to do a meeting on monday next week (7/24) that would work best for me -- otherwise I'm going to be busy moving until 8/3.

**Comment by CryptAxe** [comment]
Sorry didn't see this until just now. We'll plan on trying to meet up online some time after 8/3 then.

File: src/bitcoin-tx.cpp [hunk 1]
```diff
@@ -603,17 +603,18 @@ static void MutateTxSign(CMutableTransaction& tx, const std::string& flagStr)
         const CScript& prevPubKey = coins->vout[txin.prevout.n].scriptPubKey;
         const CAmount& amount = coins->vout[txin.prevout.n].nValue;
 
+	//TODO: This is obviously wrong, need to fix later
+	const CTransactionRef& coinbaseTx = MakeTransactionRef(tx);
```
Comment (Christewart) [inline src/bitcoin-tx.cpp:607]: Not sure what the best way to fix this is. Should we just add a parameter to `MutateTxSign()` to pass the coinbase transaction that commits to the critical hash for the `OP_BRIBEVERIFY` output we are spending?

File: src/miner.cpp [hunk 2]
```diff
@@ -433,7 +437,8 @@ CTransaction BlockAssembler::CreateSidechainWTJoinTx(uint8_t nSidechain)
 
     // Sign WT^ SCUTXO input
     const CTransaction& txToSign = mtx;
-    TransactionSignatureCreator creator(&keystoreConst, &txToSign, 0, returnAmount - amtBWT);
+    const CTransactionRef coinbaseTx = chainActive.Tip()->coinbase;
```
Comment (Christewart) [inline src/miner.cpp:440]: This coinbase transaction should be irrelevant in this function right? There is no way a `WT^` can spend an OP_BV output right?

File: src/rpc/mining.cpp [hunk 3]
```diff
@@ -543,7 +544,10 @@ UniValue getblocktemplate(const JSONRPCRequest& request)
 
         // Create new block
         CScript scriptDummy = CScript() << OP_TRUE;
-        pblocktemplate = BlockAssembler(Params()).CreateNewBlock(scriptDummy, fSupportsSegwit);
+	
+	//TODO: Come back and review how to get drivechain commitments here
+        std::vector<CTxOut> drivechainCommitments;
```
Comment (Christewart) [inline src/rpc/mining.cpp:549]: Could use help on the best way to get drivechain commitments here. Should we have some sort of specific mechanism to cache `OP_BV` outputs, and then miners can pass the `OP_BV` outputs they wish to redeem in this rpc command? Perhaps we have this function automatically look into an `OP_BV` 'wallet' and select which outputs to add to the coinbase transaction by using some sort of optimizing algorithm (back pack more than likely). Something to talk about tomorrow.

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -773,23 +774,23 @@ UniValue signrawtransaction(const JSONRPCRequest& request)
         }
         const CScript& prevPubKey = coins->vout[txin.prevout.n].scriptPubKey;
         const CAmount& amount = coins->vout[txin.prevout.n].nValue;
-
+	const CTransactionRef& coinbaseTx = chainActive.Tip()->coinbase;
```
Comment (Christewart) [inline src/rpc/rawtransaction.cpp:777]: My approach to this is wrong. There is NO way we can use `signrawtransaction` with a transaction that spends an `OP_BV` output. Only miners can create valid transactions that spend OP_BV outputs. So it is useless to call `signrawtransaction` if one of the transactions prevout scripts is a `OP_BV` output. Probably should just fail the rpc command.

File: src/rpc/mining.cpp (hunk 3 above)
Comment (CryptAxe) [reply src/rpc/mining.cpp:549]: Your idea is basically the same as what I've come up with, I'll talk to @psztorc about it tomorrow as well and get back to you :)

**Comment by psztorc** [comment]
Should we close this?

**Comment by CryptAxe** [comment]
Thank you for helping improve the design, lots of this code and ideas have been put into Drivechains

**Pull request closed without merging**